hatchery/nextflow.go

```diff
@@ -0,0 +1,1423 @@
+package hatchery
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/apparentlymart/go-cidr/cidr"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/batch"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+/*
+General TODOS:
+- Make the AWS region configurable in the hatchery config (although ideally, the user should be able to choose)
+- Make the `roleArn` configurable
+- The contents of `s3://<nextflow bucket>/<username>` are not deleted because researchers may need to keep the intermediary files.
+  We should set bucket lifecycle rules to delete after X days.
+- Can we do this long setup as a separate workspace launch step, instead of in the launch() function?
+*/
+
+// create the AWS resources required to launch nextflow workflows
+func createNextflowResources(userName string, nextflowConfig NextflowConfig) (string, string, error) {
+	var err error
+
+	// credentials and AWS services init
+	var awsConfig aws.Config
+	var awsAccountId string
+	payModel, err := getCurrentPayModel(userName)
+	if err != nil {
+		return "", "", err
+	}
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String("us-east-1"),
+	}))
+	if payModel != nil && payModel.Ecs {
+		Config.Logger.Printf("Info: pay model enabled for user '%s': creating Nextflow resources in user's AWS account", userName)
+		roleArn := fmt.Sprintf("arn:aws:iam::%s:role/csoc_adminvm", payModel.AWSAccountId)
+		awsConfig = aws.Config{
+			Credentials: stscreds.NewCredentials(sess, roleArn),
+		}
+		awsAccountId = payModel.AWSAccountId
+	} else {
+		Config.Logger.Printf("Info: pay model disabled for user '%s': creating Nextflow resources in main AWS account", userName)
+		awsConfig = aws.Config{}
+		Config.Logger.Printf("Getting AWS account ID...")
+		awsAccountId, err = getAwsAccountId(sess, &awsConfig)
+		if err != nil {
+			Config.Logger.Printf("Error getting AWS account ID: %v", err)
+			return "", "", err
+		}
```

Comment:
The access required to launch any container is at the discretion of the Data Commons maintainer and is a configuration detail. It will not be hardcoded in the code. See the feature doc: https://docs.google.com/document/d/1m0dwJsY4MxI_rytpPT5O76DZ8o-vnF2LZpFyrxfXDMk

This Nextflow doc will be updated to explain exactly how to limit access when this ^ feature is implemented

Reply:
The way this code was written and tested ensures that everything will still work in Blanket Billing, and if the limitations described in the doc are addressed we will be able to enable nextflow workspaces in Blanket Billing. But for now we will only allow that for internal devs